subsys/net/lib/lwm2m/lwm2m_obj_gateway.h

```diff
@@ -17,4 +19,9 @@
 #define LWM2M_GATEWAY_IOT_DEVICE_OBJECTS_RID 3
 /* clang-format on */
 
+typedef int (*lwm2m_engine_gateway_msg_cb)(struct lwm2m_message *msg);
+
+void lwm2m_register_gw_callback(uint16_t obj_inst_id, lwm2m_engine_gateway_msg_cb cb);
```

Comment:
```suggestion
typedef int (*lwm2m_engine_gateway_msg_cb)(struct lwm2m_message *msg);

void lwm2m_register_gw_callback(uint16_t obj_inst_id, lwm2m_engine_gateway_msg_cb cb);
```

This can now be considered as part of the public API and deserves some documentation.
It is still unclear whose responsibility is to call `lwm2m_reset_message(msg, true);` that will release the message structure.

Also, is the callback supposed to fill up the `msg` object with the response message? At least Ack is required. Please see `lwm2m_udp_receive()`.